diff --git a/modules/contrib/commerce_barion_payment/commerce_barion_payment.module b/modules/contrib/commerce_barion_payment/commerce_barion_payment.module
--- a/modules/contrib/commerce_barion_payment/commerce_barion_payment.module	
+++ b/modules/contrib/commerce_barion_payment/commerce_barion_payment.module	(date 1699871516175)
@@ -5,7 +5,7 @@
  * Contains commerce_barion_payment.module.
  */
 
-require_once DRUPAL_ROOT . '/libraries/barion-web-php/library/BarionClient.php';
+require_once DRUPAL_ROOT . '/profiles/contrib/drupaldev/libraries/barion-web-php/library/BarionClient.php';
 
 use Drupal\Core\Routing\RouteMatchInterface;
 
diff --git a/modules/contrib/commerce_barion_payment/commerce_barion_payment.install b/modules/contrib/commerce_barion_payment/commerce_barion_payment.install
--- a/modules/contrib/commerce_barion_payment/commerce_barion_payment.install	
+++ b/modules/contrib/commerce_barion_payment/commerce_barion_payment.install	(date 1699871618443)
@@ -11,8 +11,8 @@
 function commerce_barion_payment_requirements($phase) {
   $requirements = [];
   if ($phase === 'install') {
-    if (!is_dir(DRUPAL_ROOT . '/libraries/barion-web-php')
-      || !is_file(DRUPAL_ROOT . '/libraries/barion-web-php/library/BarionClient.php')) {
+    if (!is_dir(DRUPAL_ROOT . '/profiles/contrib/drupaldev/libraries/barion-web-php')
+      || !is_file(DRUPAL_ROOT . '/profiles/contrib/drupaldev/libraries/barion-web-php/library/BarionClient.php')) {
       $requirements['barion_library'] = [
         'title' => t('Barion library'),
         'description' => t('Barion library is missing, please read module documentation to learn how to include this library.'),