diff --git a/.gitlab-ci.yml b/.gitlab-ci.yml
index 90b0757b75b924f517af485c95f576f3c954fbd8..83802f1cbf179850023352442b094ce0c205cf95 100644
--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -59,6 +59,7 @@ include:
 #   OPT_IN_TEST_NEXT_MAJOR: '1'
 #   _CURL_TEMPLATES_REF: 'main'
 variables:
+  _CSPELL_WORDS: 'testmode'
   OPT_IN_TEST_PREVIOUS_MAJOR: '1'

 ###################################################################################
diff --git a/commerce_abandoned_carts.commerce_log_templates.yml b/commerce_abandoned_carts.commerce_log_templates.yml
new file mode 100644
index 0000000000000000000000000000000000000000..50c8316c42d4eb04cb4d4780cf6b20cdbaef4b49
--- /dev/null
+++ b/commerce_abandoned_carts.commerce_log_templates.yml
@@ -0,0 +1,4 @@
+mail_abandoned_cart:
+  category: commerce_order
+  label: 'Abandoned cart email'
+  template: '<p>Sent abandoned cart email to {{ to_email }}.</p>'
diff --git a/commerce_abandoned_carts.info.yml b/commerce_abandoned_carts.info.yml
index 506b2c61a42449b0168ac6f5adf7e862031faa90..1d3438adb8be2cf0b45883c67d3c4790602a9fa4 100755
--- a/commerce_abandoned_carts.info.yml
+++ b/commerce_abandoned_carts.info.yml
@@ -1,8 +1,7 @@
 name: 'Commerce Abandoned Carts'
 type: module
 description: 'Commerce Abandoned Carts'
-core: 8.x
-core_version_requirement: ^8 || ^9 || ^10
+core_version_requirement: ^9 || ^10
 package: 'Commerce'
 dependencies:
   - commerce:commerce_checkout
diff --git a/commerce_abandoned_carts.links.menu.yml b/commerce_abandoned_carts.links.menu.yml
index 48ae822c96b4a62fccc1cb6b588c66d010c50dee..a75eb1d8aed07798c1000b7eddb1ab40824be7a1 100755
--- a/commerce_abandoned_carts.links.menu.yml
+++ b/commerce_abandoned_carts.links.menu.yml
@@ -2,4 +2,4 @@ commerce_abandoned_carts.configuration:
   title: 'Abandoned carts'
   route_name: 'commerce_abandoned_carts.configuration'
   parent: 'commerce_order.configuration'
-  description: 'Configure when and how to inform the customer who abandones their shopping cart.'
+  description: 'Configure when and how to inform the customer who abandons their shopping cart.'
diff --git a/commerce_abandoned_carts.module b/commerce_abandoned_carts.module
index 5b6332152ae05c6ca724c2a26ddee2361e78c41a..68b94032b32c2baa2a5d8f89c79714cbb69c7d2c 100755
--- a/commerce_abandoned_carts.module
+++ b/commerce_abandoned_carts.module
@@ -6,8 +6,7 @@
  */

 use Drupal\Core\Routing\RouteMatchInterface;
-use Symfony\Component\Mime\Address;
-use Symfony\Component\Mime\Header\MailboxHeader;
+use Drupal\Core\Utility\Error;

 /**
  * Implements hook_help().
@@ -33,7 +32,7 @@ function commerce_abandoned_carts_theme() {
     ],
     'commerce_abandoned_carts_email' => [
       'variables' => [
-        'order' => NULL,
+        'order_entity' => NULL,
         'order_number' => NULL,
         'site_name' => NULL,
         'phone' => NULL,
@@ -46,181 +45,19 @@ function commerce_abandoned_carts_theme() {
  * Implements hook_cron().
  */
 function commerce_abandoned_carts_cron() {
-  $config = \Drupal::config('commerce_abandoned_carts.settings');
-
-  // Abort if test mode is active but no test mode email is set yet.
-  if ($config->get('testmode') && empty($config->get('testmode_email'))) {
-    \Drupal::logger('commerce_abandoned_carts')->notice('Operating in TEST mode but no test mode email has been set yet. Aborting send.');
-    return;
+  try {
+    \Drupal::service('commerce_abandoned_carts')->queueSendMails();
   }
-
-  // Get current time.
-  $time = \Drupal::time()->getRequestTime();
-  // Get timeout in seconds.
-  $carts_timeout = $config->get('timeout') * 60;
-  // Get history limit in seconds.
-  $history_limit = $config->get('history_limit') * 60;
-
-  $batch_num = 0;
-
-  // Get abandoned orders.
-  $select = \Drupal::database()->select('commerce_order', 'o');
-  $select->leftJoin('commerce_abandoned_carts', 'a', 'o.order_id = a.order_id');
-  // Only query for carts that have line items in them.
-  $select->join('commerce_order_item', 'l', 'o.order_id = l.order_id');
-  $select->addField('o', 'order_id');
-  $select->addField('o', 'mail');
-  $select->condition('o.mail', '', '!=')
-    ->isNull('a.status')
-    ->condition('o.changed', $time - $carts_timeout, '<=')
-    ->condition('o.changed', $time - $history_limit, '>=')
-    ->condition('o.state ', 'draft')
-    ->groupBy('o.order_id')
-    ->groupBy('o.mail')
-    ->orderBy('o.changed', 'ASC');
-
-  // When the Commerce Recurring module is enabled, we get draft orders from subscriptions. We can
-  // see if the draft order is linked to a subscription, and if it's not, then it's an order that
-  // should be flagged for abandonment treatment
-  if (\Drupal::moduleHandler()->moduleExists('commerce_recurring')) {
-    $select->condition('l.type', 'recurring_product_variation', '<>');
-  }
-
-  $result = $select->execute();
-
-  $row_count=$select->countQuery()->execute()->fetchField();
-
-  if ($row_count > 0) {
-    //If we'll be sending emails, set up the static elements, using local variables to avoid repeated calls...
-
-    // Set flag for the number of mails processed during this cron run.
-    $batch_limit = $config->get('batch_limit');
-
-    // Set sender.
-    if (!empty($config->get('from_email'))) {
-      $from = $message['from'] = $config->get('from_email');
-    }else{
-      $from = ''; // We will be using the Store email address
-    }
-    if (!empty($config->get('from_name'))) {
-      $from_name = $config->get('from_name');
+  catch (RuntimeException $exception) {
+    // Available since Drupal 10.1.0.
+    if (method_exists(Error::class, 'logException')) {
+      $logger = \Drupal::logger('commerce_abandoned_carts');
+      Error::logException($logger, $exception);
     }
-    else {
-      $from_name = '';  // Use the Store name
-    }
-
-    $mail_handler = \Drupal::service('commerce.mail_handler');
-    $subject = $config->get('subject');
-    $bcc_active = FALSE;
-    if ($config->get('bcc_active')) {
-      $bcc_active = TRUE;
-      $bcc_email = $config->get('bcc_email');
-    }
-
-    // Now send the mails...
-    foreach ($result as $record) {
-        // Get order.
-        $order = \Drupal::entityTypeManager()
-          ->getStorage('commerce_order')
-          ->load($record->order_id);
-
-      // Get recipient to email to. If test mode is enabled, this will be the
-      // configured testmode email address.
-      if ($config->get('testmode')) {
-        \Drupal::logger('commerce_abandoned_carts')
-          ->notice('Operating in TEST mode on order #@order', ['@order' => $order->id()]);
-        $recipient = $config->get('testmode_email');
-      }else{
-        $recipient = $record->mail;
-      }
-
-      // Get the preferred language. If the user is anonymous, this will be
-      // the site default language.
-      $customer = $order->getCustomer();
-      if ($customer && !$customer->isAnonymous()) {
-        $langcode = $customer->getPreferredLangcode();
-      }else{
-        $langcode = \Drupal::languageManager()->getDefaultLanguage()->getId();
-      }
-
-      $order_num = $order->id();
-
-      $body = [
-        '#theme'        => 'commerce_abandoned_carts_email',
-        '#order'        => $order,
-        '#order_number' => $order_num,
-        '#site_name'    => \Drupal::config('system.site')->get('name'),
-        '#phone'        => $config->get('customer_service_phone_number'),
-      ];
-
-      $params = [
-        'id'             => 'commerce_abandoned_carts',
-        'order'          => $order,
-        'langcode'       => $langcode,
-        'hookmailmodule' => 'commerce_abandoned_carts' // Use commerce_abandoned_carts_mail as hook_mail.
-      ];
-
-      // The from may depend on the Store, which is order dependent...
-      if (!$from) {
-        $from = $order->getStore()->getEmail();
-        if (!is_string($from) || strlen($from) < 1) {
-          // If the store does not have a mail address, try the site's mail address instead.
-          $from = \Drupal::config('system.site')->get('mail');
-        }
-      }
-      if (!$from_name) {
-        $from_name = $order->getStore()->name->value ?? '';
-      }
-
-      // Only set a from address if there actually is one now.
-      if (is_string($from) && strlen($from) > 0) {
-        $mailbox = new MailboxHeader('From', new Address($from, $from_name));
-        $params['from'] = $mailbox->getBodyAsString();
-      }
-
-      if ($bcc_active) {
-        $params['bcc'] = $bcc_email;
-      }
-
-      $mail_handler->sendMail($recipient, $subject, $body, $params);
-
-      // If not in test mode, flag that for this order a mail has been sent.
-      if (!$config->get('testmode')) {
-        \Drupal::database()->merge('commerce_abandoned_carts')
-          ->key(['order_id' => $record->order_id])
-          ->fields([
-            'order_id'  => $record->order_id,
-            'status'    => 1,
-            'timestamp' => \Drupal::time()->getRequestTime(),
-          ])
-          ->execute();
-      }
-
-      $batch_num++;
-
-      if ($batch_num >= $batch_limit) {
-        break;
-      }
-
+    // @todo Remove when Drupal 10 is no longer supported.
+    elseif (function_exists('watchdog_exception')) {
+      // @phpstan-ignore-next-line
+      watchdog_exception('commerce_abandoned_carts', $exception);
     }
   }
-  \Drupal::logger('commerce_abandoned_carts')->notice('Sent @num abandoned cart emails.', [
-    '@num' => $batch_num,
-  ]);
-}
-
-/**
- * Implements hook_mail().
- *
- * Prepares emails sent by the MailHandler service.
- */
-function commerce_abandoned_carts_mail_alter(&$message) {
-  if ($message['id'] != 'commerce_commerce_abandoned_carts') {
-    return;
-  }
-
-  // Make sure that 'From' is set correctly.
-  if (!empty($message['params']['from'])) {
-    $message['headers']['From'] = $message['params']['from'];
-  }
 }
diff --git a/commerce_abandoned_carts.services.yml b/commerce_abandoned_carts.services.yml
new file mode 100644
index 0000000000000000000000000000000000000000..ca3cdf68a75a2bc334af7d68ce92615049d3936f
--- /dev/null
+++ b/commerce_abandoned_carts.services.yml
@@ -0,0 +1,12 @@
+services:
+  commerce_abandoned_carts:
+    class: Drupal\commerce_abandoned_carts\AbandonedCarts
+    arguments: ['@database', '@config.factory', '@queue', '@module_handler', '@datetime.time', '@logger.channel.commerce_abandoned_carts']
+
+  commerce_abandoned_carts.abandoned_cart_mail:
+    class: Drupal\commerce_abandoned_carts\Mail\AbandonedCartMail
+    arguments: ['@database', '@commerce.mail_handler', '@config.factory', '@datetime.time']
+
+  logger.channel.commerce_abandoned_carts:
+    parent: logger.channel_base
+    arguments: ['commerce_abandoned_carts']
diff --git a/config/install/commerce_abandoned_carts.settings.yml b/config/install/commerce_abandoned_carts.settings.yml
index ed89c23c8e95bb312a5d68c0657973e2e226cd6d..236a7a7e35bd55129fd83ce77199a5c0dedadbe4 100755
--- a/config/install/commerce_abandoned_carts.settings.yml
+++ b/config/install/commerce_abandoned_carts.settings.yml
@@ -1,7 +1,5 @@
 timeout: 1440
 history_limit: 21600
-batch_limit: 5
-from_name: ''
 subject: 'Your order is incomplete.'
 customer_service_phone_number: ''
 bcc_active: false
diff --git a/config/schema/commerce_abandoned_carts.schema.yml b/config/schema/commerce_abandoned_carts.schema.yml
index 9ca79038b1a83e22b4fe19e39ebab0735998f6be..4077312b78c473b0e1d95207a7ae17759aa734d5 100755
--- a/config/schema/commerce_abandoned_carts.schema.yml
+++ b/config/schema/commerce_abandoned_carts.schema.yml
@@ -9,15 +9,6 @@ commerce_abandoned_carts.settings:
     history_limit:
       type: integer
       label: 'History limit'
-    batch_limit:
-      type: integer
-      label: 'Batch limit'
-    from_email:
-      type: email
-      label: 'From email address'
-    from_name:
-      type: string
-      label: 'From name'
     subject:
       type: string
       label: 'Subject'
diff --git a/src/AbandonedCarts.php b/src/AbandonedCarts.php
new file mode 100644
index 0000000000000000000000000000000000000000..e979246a6809fdac4b67227bbf6e5495281b40d2
--- /dev/null
+++ b/src/AbandonedCarts.php
@@ -0,0 +1,181 @@
+<?php
+
+namespace Drupal\commerce_abandoned_carts;
+
+use Drupal\Component\Datetime\TimeInterface;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Database\Connection;
+use Drupal\Core\Extension\ModuleHandlerInterface;
+use Drupal\Core\Queue\QueueFactory;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Service for sending mails to customers who abandoned their carts.
+ */
+class AbandonedCarts implements AbandonedCartsInterface {
+
+  /**
+   * The drupal database connection.
+   *
+   * @var \Drupal\Core\Database\Connection
+   */
+  protected $connection;
+
+  /**
+   * Commerce Abandoned Carts configuration.
+   *
+   * @var \Drupal\Core\Config\ImmutableConfig
+   */
+  protected $config;
+
+  /**
+   * The Commerce Abandoned Carts queue.
+   *
+   * @var \Drupal\Core\Queue\QueueInterface
+   */
+  protected $queue;
+
+  /**
+   * The module handler.
+   *
+   * @var \Drupal\Core\Extension\ModuleHandlerInterface
+   */
+  protected $moduleHandler;
+
+  /**
+   * The time service.
+   *
+   * @var \Drupal\Component\Datetime\TimeInterface
+   */
+  protected $time;
+
+  /**
+   * The logger for the commerce abandoned carts channel.
+   *
+   * @var \Psr\Log\LoggerInterface
+   */
+  protected $logger;
+
+  /**
+   * Constructs a new AbandonedCarts object.
+   *
+   * @param \Drupal\Core\Database\Connection $connection
+   *   The drupal database connection.
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   The configuration factory.
+   * @param \Drupal\Core\Queue\QueueFactory $queue_factory
+   *   The queue factory.
+   * @param \Drupal\Core\Extension\ModuleHandlerInterface $module_handler
+   *   The module handler.
+   * @param \Drupal\Component\Datetime\TimeInterface $time
+   *   The time service.
+   * @param \Psr\Log\LoggerInterface $logger
+   *   The logger for the commerce abandoned carts channel.
+   */
+  public function __construct(Connection $connection, ConfigFactoryInterface $config_factory, QueueFactory $queue_factory, ModuleHandlerInterface $module_handler, TimeInterface $time, LoggerInterface $logger) {
+    $this->connection = $connection;
+    $this->config = $config_factory->get('commerce_abandoned_carts.settings');
+    $this->queue = $queue_factory->get('commerce_abandoned_carts_send_mail');
+    $this->moduleHandler = $module_handler;
+    $this->time = $time;
+    $this->logger = $logger;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAbandonedOrders($limit = NULL) {
+    // Get current time.
+    $time = $this->time->getRequestTime();
+    // Get timeout in seconds.
+    $carts_timeout = $this->config->get('timeout') * 60;
+    // Get history limit in seconds.
+    $history_limit = $this->config->get('history_limit') * 60;
+
+    $select = $this->connection->select('commerce_order', 'o');
+    $select->leftJoin('commerce_abandoned_carts', 'a', 'o.order_id = a.order_id');
+    // Only query for carts that have line items in them.
+    $select->join('commerce_order_item', 'l', 'o.order_id = l.order_id');
+    $select->addField('o', 'order_id');
+    $select->addField('o', 'mail');
+    $select->condition('o.mail', '', '!=')
+      ->isNull('a.status')
+      ->condition('o.changed', $time - $carts_timeout, '<=')
+      ->condition('o.changed', $time - $history_limit, '>=')
+      ->condition('o.state ', 'draft')
+      ->groupBy('o.order_id')
+      ->groupBy('o.mail')
+      ->orderBy('o.changed', 'ASC');
+
+    if ($limit) {
+      $select->range(0, $limit);
+    }
+
+    // When the Commerce Recurring module is enabled, we get draft orders from
+    // subscriptions. We can see if the draft order is linked to a subscription,
+    // and if it's not, then it's an order that should be flagged for
+    // abandonment treatment.
+    if ($this->moduleHandler->moduleExists('commerce_recurring')) {
+      $select->condition('l.type', 'recurring_product_variation', '<>');
+    }
+
+    return $select->execute()
+      ->fetchAll();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function queueSendMails($limit = 100) {
+    if ($this->config->get('testmode')) {
+      if (empty($this->config->get('testmode_email'))) {
+        throw new \RuntimeException('Operating in TEST mode but no test mode email has been set yet. Aborting send.');
+      }
+    }
+
+    $params = [];
+
+    // Optionally set BCC.
+    if ($this->config->get('bcc_active') && $this->config->get('bcc_email')) {
+      $params['bcc'] = $this->config->get('bcc_email');
+    }
+
+    // Queue sending abandoned cart mails.
+    $order_ids = [];
+    foreach ($this->getAbandonedOrders($limit) as $record) {
+      $order_ids[] = $record->order_id;
+
+      // Get recipient to email to. If test mode is enabled, this will be the
+      // configured testmode email address.
+      if ($this->config->get('testmode')) {
+        $recipient = $this->config->get('testmode_email');
+      }
+      else {
+        $recipient = $record->mail;
+      }
+
+      // Put item on the queue.
+      $this->queue->createItem([$record->order_id, $recipient, $params]);
+
+      // If not in test mode, flag that for this order a mail has been queued.
+      if (!$this->config->get('testmode')) {
+        $this->connection->merge('commerce_abandoned_carts')
+          ->keys(['order_id' => $record->order_id])
+          ->fields([
+            'order_id' => $record->order_id,
+            'status' => static::QUEUED,
+            'timestamp' => $this->time->getRequestTime(),
+          ])
+          ->execute();
+      }
+    }
+
+    // Log for which orders testmode was used.
+    if ($this->config->get('testmode')) {
+      $this->logger->notice('Operating in TEST mode on orders @orders', [
+        '@orders' => implode(', ', $order_ids),
+      ]);
+    }
+  }
+
+}
diff --git a/src/AbandonedCartsInterface.php b/src/AbandonedCartsInterface.php
new file mode 100644
index 0000000000000000000000000000000000000000..b36388ad895b119fe297cb10df2e20048442b120
--- /dev/null
+++ b/src/AbandonedCartsInterface.php
@@ -0,0 +1,47 @@
+<?php
+
+namespace Drupal\commerce_abandoned_carts;
+
+/**
+ * Interface for a service for sending mails to users who abandoned their cart.
+ */
+interface AbandonedCartsInterface {
+
+  /**
+   * Flag that a particular order has been queued for sending.
+   *
+   * @var int
+   */
+  const QUEUED = 1;
+
+  /**
+   * Flag that a mail for an order was sent.
+   *
+   * @var int
+   */
+  const SENT = 2;
+
+  /**
+   * Returns a list of abandoned orders.
+   *
+   * @param int $limit
+   *   The maximum number of orders to return.
+   *
+   * @return object[]
+   *   A list of objects with the following properties:
+   *   - order_id
+   *     The order ID.
+   *   - mail
+   *     The mail address to send an abandoned cart mail to.
+   */
+  public function getAbandonedOrders($limit = NULL);
+
+  /**
+   * Queues sending mails.
+   *
+   * @param int $limit
+   *   (optional) The limit of many mails to queue. Defaults to 100.
+   */
+  public function queueSendMails($limit = 100);
+
+}
diff --git a/src/Form/AdminForm.php b/src/Form/AdminForm.php
index 39a0c852be3bf40369e0c9d79c23c52d00512837..293cb3b59bd907ff2e8a943af9c96270eda61b97 100755
--- a/src/Form/AdminForm.php
+++ b/src/Form/AdminForm.php
@@ -49,42 +49,6 @@ class AdminForm extends ConfigFormBase {
       '#description' => $this->t('What is the limit (in minutes) to how far back to search for abandoned carts. Default is 15 days.'),
     ];

-    // @todo replace with queue.
-    $options = [
-      5 => '5',
-      10 => '10',
-      25 => '25',
-      50 => '50',
-      75 => '75',
-      100 => '100',
-    ];
-
-    $form['batch_limit'] = [
-      '#type' => 'select',
-      '#title' => $this->t('Batch limit'),
-      '#options' => $options,
-      '#default_value' => $config->get('batch_limit'),
-      '#description' => $this->t('What is the maximum emails to send per cron run? Note, larger batches may cause performance issues.'),
-    ];
-
-    $form['from_email'] = [
-      '#type' => 'email',
-      '#title' => $this->t('From email address'),
-      '#default_value' => $config->get('from_email'),
-      '#size' => 60,
-      '#maxlength' => 128,
-      '#description' => $this->t('Enter the email address to send the emails from. Leave blank to use the store email address.'),
-    ];
-
-    $form['from_name'] = [
-      '#type' => 'textfield',
-      '#title' => $this->t('From name'),
-      '#default_value' => $config->get('from_name'),
-      '#size' => 60,
-      '#maxlength' => 128,
-      '#description' => $this->t('Enter the name to send the emails from.  Leave blank to use site name.'),
-    ];
-
     $form['subject'] = [
       '#type' => 'textfield',
       '#title' => $this->t('Subject'),
@@ -100,7 +64,7 @@ class AdminForm extends ConfigFormBase {
       '#default_value' => $config->get('customer_service_phone_number'),
       '#size' => 60,
       '#maxlength' => 128,
-      '#description' => $this->t('Enter a phone number to be displayed in the email template for customers who may have had trouble checking out. Leave empty to obmit from email.'),
+      '#description' => $this->t('Enter a phone number to be displayed in the email template for customers who may have had trouble checking out. Leave empty to omit it from the email message.'),
     ];

     $form['bcc_active'] = [
@@ -179,9 +143,6 @@ class AdminForm extends ConfigFormBase {
     $this->config('commerce_abandoned_carts.settings')
       ->set('timeout', $form_state->getValue('timeout'))
       ->set('history_limit', $form_state->getValue('history_limit'))
-      ->set('batch_limit', $form_state->getValue('batch_limit'))
-      ->set('from_email', $form_state->getValue('from_email'))
-      ->set('from_name', $form_state->getValue('from_name'))
       ->set('subject', $form_state->getValue('subject'))
       ->set('customer_service_phone_number', $form_state->getValue('customer_service_phone_number'))
       ->set('bcc_active', $form_state->getValue('bcc_active'))
diff --git a/src/Mail/AbandonedCartMail.php b/src/Mail/AbandonedCartMail.php
new file mode 100644
index 0000000000000000000000000000000000000000..be3711032fa200095b9626ac8b8984b3a394ea69
--- /dev/null
+++ b/src/Mail/AbandonedCartMail.php
@@ -0,0 +1,162 @@
+<?php
+
+namespace Drupal\commerce_abandoned_carts\Mail;
+
+use Drupal\commerce\MailHandlerInterface;
+use Drupal\commerce_abandoned_carts\AbandonedCartsInterface;
+use Drupal\commerce_order\Entity\OrderInterface;
+use Drupal\Component\Datetime\TimeInterface;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Database\Connection;
+use Symfony\Component\Mime\Address;
+use Symfony\Component\Mime\Header\MailboxHeader;
+
+/**
+ * Service for sending an abandoned cart mail.
+ */
+class AbandonedCartMail implements AbandonedCartMailInterface {
+
+  /**
+   * The drupal database connection.
+   *
+   * @var \Drupal\Core\Database\Connection
+   */
+  protected $connection;
+
+  /**
+   * The mail handler.
+   *
+   * @var \Drupal\commerce\MailHandlerInterface
+   */
+  protected $mailHandler;
+
+  /**
+   * Commerce Abandoned Carts configuration.
+   *
+   * @var \Drupal\Core\Config\ImmutableConfig
+   */
+  protected $config;
+
+  /**
+   * Site configuration.
+   *
+   * @var \Drupal\Core\Config\ImmutableConfig
+   */
+  protected $siteConfig;
+
+  /**
+   * The time service.
+   *
+   * @var \Drupal\Component\Datetime\TimeInterface
+   */
+  protected $time;
+
+  /**
+   * Constructs a new AbandonedCartMail object.
+   *
+   * @param \Drupal\Core\Database\Connection $connection
+   *   The drupal database connection.
+   * @param \Drupal\commerce\MailHandlerInterface $mail_handler
+   *   The mail handler.
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   The configuration factory.
+   * @param \Drupal\Component\Datetime\TimeInterface $time
+   *   The time service.
+   */
+  public function __construct(Connection $connection, MailHandlerInterface $mail_handler, ConfigFactoryInterface $config_factory, TimeInterface $time) {
+    $this->connection = $connection;
+    $this->mailHandler = $mail_handler;
+    $this->config = $config_factory->get('commerce_abandoned_carts.settings');
+    $this->siteConfig = $config_factory->get('system.site');
+    $this->time = $time;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function send(OrderInterface $order, $to = NULL, array $params = []) {
+    $to ?? $order->getEmail();
+    if (!$to) {
+      // The recipient email should not be empty.
+      return FALSE;
+    }
+
+    // Setup mail parameters.
+    $params += [
+      'id' => 'abandoned_cart',
+      'order' => $order,
+    ];
+
+    // Determine the language to use, if not yet set.
+    if (!isset($params['langcode'])) {
+      $customer = $order->getCustomer();
+      if ($customer->isAuthenticated()) {
+        $params['langcode'] = $customer->getPreferredLangcode();
+      }
+    }
+
+    // Setup mail body.
+    $body = [
+      '#theme' => 'commerce_abandoned_carts_email',
+      '#order_entity' => $order,
+      '#order_number' => $order->id(),
+      '#site_name' => $this->siteConfig->get('name'),
+      '#phone' => $this->config->get('customer_service_phone_number'),
+    ];
+
+    // Determine the from address.
+    $from = $this->getFromAddress($order);
+    if (is_string($from)) {
+      $params['from'] = $from;
+    }
+
+    // And send the mail!
+    $success = $this->mailHandler->sendMail($to, $this->config->get('subject'), $body, $params);
+
+    // If not in test mode, flag that for this order a mail has been sent.
+    if ($success && !$this->config->get('testmode')) {
+      $this->connection->merge('commerce_abandoned_carts')
+        ->keys(['order_id' => $order->id()])
+        ->fields([
+          'order_id' => $order->id(),
+          'status' => AbandonedCartsInterface::SENT,
+          'timestamp' => $this->time->getRequestTime(),
+        ])
+        ->execute();
+    }
+
+    return $success;
+  }
+
+  /**
+   * Tries to resolve the from address.
+   *
+   * @param \Drupal\commerce_order\Entity\OrderInterface $order
+   *   The order.
+   *
+   * @return string|null
+   *   A string when a from address could be resolved. Null otherwise.
+   */
+  protected function getFromAddress(OrderInterface $order): ?string {
+    /** @var \Drupal\commerce_store\Entity\StoreInterface $store */
+    $store = $order->getStore();
+
+    $from = $store->getEmail();
+    $from_name = $store->getName() ?? '';
+    if (!is_string($from) || strlen($from) < 1) {
+      // If the store does not have a mail address, try the site's mail address
+      // instead.
+      $from = $this->siteConfig->get('mail');
+    }
+
+    // Only set a from address if there actually is one now.
+    if (is_string($from) && strlen($from) > 0) {
+      $mailbox = new MailboxHeader('From', new Address($from, $from_name));
+      return $mailbox->getBodyAsString();
+    }
+
+    // In other cases, return nothing.
+    return NULL;
+  }
+
+}
diff --git a/src/Mail/AbandonedCartMailInterface.php b/src/Mail/AbandonedCartMailInterface.php
new file mode 100644
index 0000000000000000000000000000000000000000..5e5b3c537533592d1fd0c74a15465d1cf24107ab
--- /dev/null
+++ b/src/Mail/AbandonedCartMailInterface.php
@@ -0,0 +1,28 @@
+<?php
+
+namespace Drupal\commerce_abandoned_carts\Mail;
+
+use Drupal\commerce_order\Entity\OrderInterface;
+
+/**
+ * Interface for abandoned cart mail service.
+ */
+interface AbandonedCartMailInterface {
+
+  /**
+   * Sends the abandoned cart email.
+   *
+   * @param \Drupal\commerce_order\Entity\OrderInterface $order
+   *   The order.
+   * @param string $to
+   *   The address the email will be sent to. Must comply with RFC 2822.
+   *   Defaults to the order email.
+   * @param array $params
+   *   (optional) Additional params to set for the mail.
+   *
+   * @return bool
+   *   TRUE if the email was sent successfully, FALSE otherwise.
+   */
+  public function send(OrderInterface $order, $to = NULL, array $params = []);
+
+}
diff --git a/src/Plugin/QueueWorker/AbandonedCartMail.php b/src/Plugin/QueueWorker/AbandonedCartMail.php
new file mode 100644
index 0000000000000000000000000000000000000000..5fdecccad9b698941fed6477d64b060ea29fec16
--- /dev/null
+++ b/src/Plugin/QueueWorker/AbandonedCartMail.php
@@ -0,0 +1,81 @@
+<?php
+
+namespace Drupal\commerce_abandoned_carts\Plugin\QueueWorker;
+
+use Drupal\commerce_abandoned_carts\Mail\AbandonedCartMailInterface;
+use Drupal\commerce_order\OrderStorage;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\Core\Queue\QueueWorkerBase;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * A queue worker for sending mails to people who abandoned their cart.
+ *
+ * @QueueWorker(
+ *   id = "commerce_abandoned_carts_send_mail",
+ *   title = @Translation("Abandoned carts mails sending"),
+ *   cron = {"time" = 60}
+ * )
+ */
+class AbandonedCartMail extends QueueWorkerBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The abandoned_carts service.
+   *
+   * @var \Drupal\commerce_abandoned_carts\Mail\AbandonedCartsInterface
+   */
+  protected $abandonedCartMail;
+
+  /**
+   * The order storage.
+   *
+   * @var \Drupal\commerce_order\OrderStorage
+   */
+  protected $orderStorage;
+
+  /**
+   * Constructs a new AbandonedCartMail object.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param array $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\commerce_abandoned_carts\Mail\AbandonedCartMailInterface $abandoned_cart_mail
+   *   The abandoned_cart mail service.
+   * @param \Drupal\commerce_order\OrderStorage $order_storage
+   *   The order storage.
+   */
+  public function __construct(array $configuration, $plugin_id, array $plugin_definition, AbandonedCartMailInterface $abandoned_cart_mail, OrderStorage $order_storage) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->abandonedCartMail = $abandoned_cart_mail;
+    $this->orderStorage = $order_storage;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('commerce_abandoned_carts.abandoned_cart_mail'),
+      $container->get('entity_type.manager')->getStorage('commerce_order')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function processItem($data) {
+    [$order_id, $recipient, $params] = $data;
+
+    $order = $this->orderStorage->load($order_id);
+    if ($order) {
+      $this->abandonedCartMail->send($order, $recipient, $params);
+    }
+  }
+
+}
diff --git a/templates/commerce-abandoned-carts-email.html.twig b/templates/commerce-abandoned-carts-email.html.twig
index 262b75db223181a4c588d96edd27ec6b15f87e52..aadb3b2faf854e36dd97c6c95cb076d68526a6c5 100755
--- a/templates/commerce-abandoned-carts-email.html.twig
+++ b/templates/commerce-abandoned-carts-email.html.twig
@@ -4,7 +4,7 @@
  * Template for the mail to the customer who did not finish their order.
  *
  * Available variables:
- * - order: The order entity.
+ * - order_entity: The order entity.
  * - order_number: The order number.
  * - site_name: The name of the website.
  * - phone: The customer service phone number, configured on the abandoned carts
@@ -15,7 +15,7 @@
 #}
 Dear valued customer,<br />

-<p>This is a reminder that <a href="{{ url('<front>') }}/cart">your order (#{{ order_number }}) at {{ site_name }}</a> is incomplete. Please note that you have <strong>not</strong> been charged.</p>
+<p>This is a reminder that <a href="{{ url('<front>') }}cart">your order (#{{ order_number }}) at {{ site_name }}</a> is incomplete. Please note that you have <strong>not</strong> been charged.</p>
 <p>You may return to the website to complete your order at any time, but the link given above will only work in the browser that you originally used, when you started the order.</p>

 {% if phone %}
@@ -23,7 +23,6 @@ Dear valued customer,<br />
 {% else %}
   <p>If you've experienced any trouble with your checkout, or if you have any questions, please contact us for assistance.</p>
 {% endif %}
-<br />

 Sincerely,<br />
 <br />
diff --git a/tests/src/Functional/Form/AdminFormTest.php b/tests/src/Functional/Form/AdminFormTest.php
index 625b77259dd42a0c7947dbf773102af1bcffa863..5b0369b4fce83b3a2eb4ff5cee8b76fbaf4608ad 100755
--- a/tests/src/Functional/Form/AdminFormTest.php
+++ b/tests/src/Functional/Form/AdminFormTest.php
@@ -22,7 +22,7 @@ class AdminFormTest extends BrowserTestBase {
   protected static $modules = ['commerce_abandoned_carts'];

   /**
-   * Tests that an user without the right permissions cannot access the form.
+   * Tests that a user without the right permissions cannot access the form.
    */
   public function testAccessDenied() {
     $account = $this->drupalCreateUser(['access content']);
@@ -44,9 +44,6 @@ class AdminFormTest extends BrowserTestBase {
     $edit = [
       'timeout' => 120,
       'history_limit' => 10080,
-      'batch_limit' => 10,
-      'from_email' => 'fromme@example.com',
-      'from_name' => 'From me',
       'subject' => 'You did not complete your order',
       'customer_service_phone_number' => '123',
       'bcc_active' => 1,
@@ -61,9 +58,6 @@ class AdminFormTest extends BrowserTestBase {
     $expected = [
       'timeout' => 120,
       'history_limit' => 10080,
-      'batch_limit' => 10,
-      'from_email' => 'fromme@example.com',
-      'from_name' => 'From me',
       'subject' => 'You did not complete your order',
       'customer_service_phone_number' => '123',
       'bcc_active' => 1,
diff --git a/tests/src/Functional/MailTest.php b/tests/src/Functional/MailTest.php
index 4e56290299ddea81b2423899fdc2796b025b23df..2fe73939c4beddeb9a459db3e5c86210a5dea0d4 100755
--- a/tests/src/Functional/MailTest.php
+++ b/tests/src/Functional/MailTest.php
@@ -187,7 +187,7 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert that the customer received an email.
-    $mails = $this->getMails(['key' => 'commerce_abandoned_carts']);
+    $mails = $this->getMails(['key' => 'abandoned_cart']);
     $this->assertEquals('guest@example.com', $mails[0]['to']);

     // Run cron again to ensure that the customer only receives one mail.
@@ -214,7 +214,7 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert that the customer received an email.
-    $mails = $this->getMails(['key' => 'commerce_abandoned_carts']);
+    $mails = $this->getMails(['key' => 'abandoned_cart']);
     $this->assertEquals('customer@example.com', $mails[0]['to']);

     // Run cron again to ensure that the customer only receives one mail.
@@ -244,7 +244,7 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert that the customer received an email.
-    $mails = $this->getMails(['key' => 'commerce_abandoned_carts']);
+    $mails = $this->getMails(['key' => 'abandoned_cart']);
     $this->assertEquals('customer@example.com', $mails[0]['to']);

     // Run cron again to ensure that the customer only receives one mail.
@@ -278,7 +278,7 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert no mails sent.
-    $this->assertEmpty($this->getMails(['key' => 'commerce_abandoned_carts']), 'No emails have been sent.');
+    $this->assertEmpty($this->getMails(['key' => 'abandoned_cart']), 'No emails have been sent.');
   }

   /**
@@ -300,7 +300,7 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert no mails sent.
-    $this->assertEmpty($this->getMails(['key' => 'commerce_abandoned_carts']), 'No emails have been sent.');
+    $this->assertEmpty($this->getMails(['key' => 'abandoned_cart']), 'No emails have been sent.');
   }

   /**
@@ -324,38 +324,10 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert that the customer received an email.
-    $mails = $this->getMails(['key' => 'commerce_abandoned_carts']);
+    $mails = $this->getMails(['key' => 'abandoned_cart']);
     $this->assertEquals('customer@example.com', $mails[0]['to']);
   }

-  /**
-   * Tests sending mail with a custom from name and mail set.
-   */
-  public function testWithFrom() {
-    $this->config('commerce_abandoned_carts.settings')
-      ->set('from_email', 'foo@example.com')
-      ->set('from_name', 'Foo, Bar, and Baz')
-      ->save();
-
-    // Customer adds a product to the cart, goes to checkout, but then stops.
-    $account = $this->drupalCreateUser(['access content'], 'customer');
-    $this->drupalLogin($account);
-    $this->addProductToCart($this->product);
-    $this->goToCheckout();
-
-    // Manually set the order's last update date to a day ago.
-    $order = Order::load(1);
-    $this->setOrderChangedDate(1, $order->changed->value - static::SECONDS_IN_DAY);
-
-    // Run cron.
-    $this->cronRun();
-
-    // Assert that the from name was used in the mail.
-    $mails = $this->getMails(['key' => 'commerce_abandoned_carts']);
-    file_put_contents('/tmp/aba.txt', print_r($mails[0]['headers'], TRUE));
-    $this->assertEquals('"Foo, Bar, and Baz" <foo@example.com>', $mails[0]['headers']['From']);
-  }
-
   /**
    * Tests with BCC enabled.
    */
@@ -380,7 +352,7 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert that mail was sent bcc.
-    $mails = $this->getMails(['key' => 'commerce_abandoned_carts']);
+    $mails = $this->getMails(['key' => 'abandoned_cart']);
     $this->assertEquals('bcc@example.com', $mails[0]['headers']['Bcc']);

     // Now deactivate bcc.
@@ -388,7 +360,7 @@ class MailTest extends CommerceBrowserTestBase {
       ->set('bcc_active', FALSE)
       ->save();

-    // An other customer abandones their cart.
+    // An other customer abandons their cart.
     $account = $this->drupalCreateUser(['access content'], 'customer2');
     $this->drupalLogin($account);
     $this->addProductToCart($this->product);
@@ -402,7 +374,7 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert that mail was *not* sent bcc.
-    $mails = $this->getMails(['key' => 'commerce_abandoned_carts']);
+    $mails = $this->getMails(['key' => 'abandoned_cart']);
     $this->assertTrue(empty($mails[1]['headers']['Bcc']));
   }

@@ -429,7 +401,7 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert that the mail was sent to the test mail address only.
-    $mails = $this->getMails(['key' => 'commerce_abandoned_carts']);
+    $mails = $this->getMails(['key' => 'abandoned_cart']);
     $this->assertEquals('test@example.com', $mails[0]['to']);
     $this->assertCount(1, $mails);

@@ -442,7 +414,7 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert that the customer now received a mail.
-    $mails = $this->getMails(['key' => 'commerce_abandoned_carts']);
+    $mails = $this->getMails(['key' => 'abandoned_cart']);
     $this->assertEquals('customer@example.com', $mails[1]['to']);
     $this->assertCount(2, $mails);
   }
@@ -469,7 +441,7 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert no mails sent, since there was no mail configured.
-    $this->assertEmpty($this->getMails(['key' => 'commerce_abandoned_carts']), 'No emails have been sent.');
+    $this->assertEmpty($this->getMails(['key' => 'abandoned_cart']), 'No emails have been sent.');
   }

   /**
@@ -500,14 +472,14 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert that the customer received an email in French.
-    $mails = $this->getMails(['key' => 'commerce_abandoned_carts']);
+    $mails = $this->getMails(['key' => 'abandoned_cart']);
     $this->assertEquals('fr', $mails[0]['langcode']);
   }

   /**
-   * Tests if the mail gets sent user's prefered language.
+   * Tests if the mail gets sent user's preferred language.
    *
-   * The mail can only get sent in the user's prefered language if the user is
+   * The mail can only get sent in the user's preferred language if the user is
    * authenticated.
    */
   public function testMailLanguageAuthenticatedCustomer() {
@@ -540,7 +512,7 @@ class MailTest extends CommerceBrowserTestBase {
     $this->cronRun();

     // Assert that the customer received an email in Dutch.
-    $mails = $this->getMails(['key' => 'commerce_abandoned_carts']);
+    $mails = $this->getMails(['key' => 'abandoned_cart']);
     $this->assertEquals('nl', $mails[0]['langcode']);
   }

diff --git a/tests/src/Kernel/MailTest.php b/tests/src/Kernel/Mail/AbandonedCartMailTest.php
similarity index 58%
rename from tests/src/Kernel/MailTest.php
rename to tests/src/Kernel/Mail/AbandonedCartMailTest.php
index e8333e36b27c16b5ba1f9aa44b16bc88ece8627c..a60cdd304e75fe7957788184c9f30c717ef30674 100755
--- a/tests/src/Kernel/MailTest.php
+++ b/tests/src/Kernel/Mail/AbandonedCartMailTest.php
@@ -1,6 +1,6 @@
 <?php

-namespace Drupal\Tests\commerce_abandoned_carts\Kernel;
+namespace Drupal\Tests\commerce_abandoned_carts\Kernel\Mail;

 use Drupal\commerce_order\Entity\Order;
 use Drupal\Core\Test\AssertMailTrait;
@@ -8,11 +8,11 @@ use Drupal\profile\Entity\Profile;
 use Drupal\Tests\commerce\Kernel\CommerceKernelTestBase;

 /**
- * Tests that customers that abandoned their carts receive a mail.
+ * @coversDefaultClass \Drupal\commerce_abandoned_carts\Mail\AbandonedCartMail
  *
  * @group commerce_abandoned_carts
  */
-class MailTest extends CommerceKernelTestBase {
+class AbandonedCartMailTest extends CommerceKernelTestBase {

   use AssertMailTrait;

@@ -94,59 +94,21 @@ class MailTest extends CommerceKernelTestBase {
   }

   /**
-   * Tests if the from setting is respected.
+   * Tests sending an email.
    *
-   * @param string $expected_from
-   *   The expected "From" header.
-   * @param array $settings
-   *   Commerce abandoned cart settings.
-   *
-   * @dataProvider providerFrom
+   * @covers ::send
    */
-  public function testFrom($expected_from, array $settings) {
-    $config = $this->config('commerce_abandoned_carts.settings');
-    foreach ($settings as $key => $value) {
-      $config->set($key, $value);
-    }
-    $config->save();
-
-    $this->markTestIncomplete('Needs to use mail handler');
-
-    $params = [
-      'order' => $this->order,
-    ];
-    $this->container->get('plugin.manager.mail')->mail('commerce_abandoned_carts', 'abandoned_cart', 'test@example.com', 'en', $params, NULL, TRUE);
+  public function testSend() {
+    $this->container->get('commerce_abandoned_carts.abandoned_cart_mail')->send($this->order, 'test@example.com');

     $emails = $this->getMails();
     $email = reset($emails);
-    $this->assertEquals($expected_from, $email['headers']['From']);
-  }

-  /**
-   * Data provider for ::testFrom().
-   */
-  public function providerFrom() {
-    return [
-      'default' => [
-        'Default store <admin@example.com>',
-        [],
-      ],
-      'from_email' => [
-        'Default store <commerce_abandoned_carts@example.com>',
-        ['from_email' => 'commerce_abandoned_carts@example.com'],
-      ],
-      'from_name' => [
-        'Bar <admin@example.com>',
-        ['from_name' => 'Bar'],
-      ],
-      'from_both' => [
-        'Bar <commerce_abandoned_carts@example.com>',
-        [
-          'from_email' => 'commerce_abandoned_carts@example.com',
-          'from_name' => 'Bar',
-        ],
-      ],
-    ];
+    $this->assertEquals('commerce_abandoned_cart', $email['id']);
+    $this->assertEquals('test@example.com', $email['to']);
+    $this->assertEquals('Default store <admin@example.com>', $email['from']);
+    $this->assertEquals('Your order is incomplete.', $email['subject']);
+    $this->assertStringContainsString('This is a reminder that your order (#1) at Drupal [1] is incomplete.', $email['body']);
   }

 }
