diff --git a/modules/contrib/commerce/modules/price/src/Price.php b/modules/contrib/commerce/modules/price/src/Price.php
--- a/modules/contrib/commerce/modules/price/src/Price.php	
+++ b/modules/contrib/commerce/modules/price/src/Price.php	(date 1707202783925)
@@ -313,7 +313,7 @@
    */
   protected function assertSameCurrency(Price $first_price, Price $second_price) {
     if ($first_price->getCurrencyCode() != $second_price->getCurrencyCode()) {
-      throw new CurrencyMismatchException(sprintf('The provided prices have mismatched currencies: %s, %s.', $first_price->__toString(), $second_price->__toString()));
+      //throw new CurrencyMismatchException(sprintf('The provided prices have mismatched currencies: %s, %s.', $first_price->__toString(), $second_price->__toString()));
     }
   }
 
