diff --git a/facebook_pixel.module b/facebook_pixel.module
index cb7c877..79f9ad8 100755
--- a/facebook_pixel.module
+++ b/facebook_pixel.module
@@ -10,6 +10,7 @@ use Drupal\Core\Entity\EntityInterface;
 use Drupal\Core\Entity\Display\EntityViewDisplayInterface;
 use Drupal\Component\Utility\Xss;
 use Drupal\Core\Config\ImmutableConfig;
+use Drupal\user\Entity\User;
 
 /**
  * Implements hook_help().
@@ -46,9 +47,19 @@ function facebook_pixel_page_attachments(array &$page) {
 
   $events = [];
   foreach ($facebook_event->getEvents() as $event) {
+    $user = User::load(\Drupal::currentUser()->id());
+    $event['eventID'] =  $user->uuid();
+
+    if ($event['event'] == 'Purchase') {
+      $order = \Drupal::request()->attributes->get('commerce_order');
+      $order_id = $order->id();
+      $event['eventID'] = 'checkout.purchase.' . $order_id;
+    }
+
     $events[] = [
       'event' => Xss::filter($event['event']),
       'data' => Xss::filter(json_encode($event['data'])),
+      'eventID' => !empty($event['eventID']) ? Xss::filter(json_encode($event['eventID'])) : "",
     ];
   }
 
diff --git a/js/facebook_pixel.js b/js/facebook_pixel.js
index 6178b01..12fcdcc 100644
--- a/js/facebook_pixel.js
+++ b/js/facebook_pixel.js
@@ -72,7 +72,7 @@
         fbq('track', 'PageView');
 
         drupalSettings.facebook_pixel.events.forEach(function (event) {
-          fbq("track", event['event'], JSON.parse(event['data']));
+          fbq("track", event['event'], event['data'], {'eventID': event['eventID']});
         });
       });
     }
