diff --git a/modules/contrib/fivestar/js/fivestar.js b/modules/contrib/fivestar/js/fivestar.js
--- a/modules/contrib/fivestar/js/fivestar.js	
+++ b/modules/contrib/fivestar/js/fivestar.js	(date 1696932697159)
@@ -48,7 +48,7 @@
         $container.addClass('fivestar-widget-' + $options.length);
         $container.find('a').bind('click', $this, Drupal.behaviors.fivestar.rate).bind('mouseover', $this, Drupal.behaviors.fivestar.hover);
         $container.bind('mouseover mouseout', $this, Drupal.behaviors.fivestar.hover);
-        $select.after($container).css('display', 'none');
+        $select.after($container).css('visibility', 'hidden');
       });
     },
     rate: function(event) {
diff --git a/modules/contrib/fivestar/src/Element/Fivestar.php b/modules/contrib/fivestar/src/Element/Fivestar.php
--- a/modules/contrib/fivestar/src/Element/Fivestar.php	
+++ b/modules/contrib/fivestar/src/Element/Fivestar.php	(date 1696932365298)
@@ -67,7 +67,7 @@
       $title = $complete_form['#node']->title;
     }
 
-    $options = ['-' => t('Select rating')];
+    $options = ['' => t('Select rating')];
     for ($i = 1; $i <= $element['#stars']; $i++) {
       $this_value = ceil($i * 100 / $element['#stars']);
       $options[$this_value] = t('Give @title @star/@count', [
diff --git a/modules/contrib/fivestar/src/Element/Fivestar.php b/modules/contrib/fivestar/src/Element/Fivestar.php
--- a/modules/contrib/fivestar/src/Element/Fivestar.php
+++ b/modules/contrib/fivestar/src/Element/Fivestar.php	(date 1697008511711)
@@ -85,6 +85,7 @@
     if (self::userCanVote($element)) {
       $element['vote'] = [
         '#type' => 'select',
+        '#title' => $element['#title'],
         '#description' => self::getElementDescription($element),
         '#options' => $options,
         '#rating' => $values['vote_average'],
diff --git a/modules/contrib/fivestar/src/Plugin/Field/FieldWidget/StarsWidget.php b/modules/contrib/fivestar/src/Plugin/Field/FieldWidget/StarsWidget.php
--- a/modules/contrib/fivestar/src/Plugin/Field/FieldWidget/StarsWidget.php	
+++ b/modules/contrib/fivestar/src/Plugin/Field/FieldWidget/StarsWidget.php	(date 1696933093573)
@@ -130,6 +130,7 @@
       '#widget' => $display_settings,
       '#settings' => $display_settings,
       '#show_static_result' => !$voting_is_allowed,
+      '#required' => $element['#required'],
     ];
 
     return $element;
diff --git a/modules/contrib/fivestar/src/Plugin/Field/FieldWidget/SelectWidget.php b/modules/contrib/fivestar/src/Plugin/Field/FieldWidget/SelectWidget.php
--- a/modules/contrib/fivestar/src/Plugin/Field/FieldWidget/SelectWidget.php	
+++ b/modules/contrib/fivestar/src/Plugin/Field/FieldWidget/SelectWidget.php	(date 1696933205393)
@@ -41,7 +41,7 @@
       '#type' => 'select',
       '#title' => $element['#title'],
       '#empty_option' => $this->t('Select rating:'),
-      '#empty_value' => '-',
+      '#empty_value' => '',
       '#options' => $options,
       '#required' => $items[$delta]->getFieldDefinition()->isRequired(),
       '#default_value' => $items[$delta]->rating ?? 0,
