Index: web/modules/contrib/terms_of_use/terms_of_use.module
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/web/modules/contrib/terms_of_use/terms_of_use.module b/web/modules/contrib/terms_of_use/terms_of_use.module
--- a/web/modules/contrib/terms_of_use/terms_of_use.module	
+++ b/web/modules/contrib/terms_of_use/terms_of_use.module	(date 1709112209597)
@@ -45,7 +45,7 @@
   }
   $data = '';
 
-  $checkbox_title = $config->get('terms_of_use_label_checkbox');
+  $checkbox_title = t($config->get('terms_of_use_label_checkbox'));
   if (!empty($node)) {
     if (!empty($checkbox_title) && strpos($checkbox_title, '@link') !== FALSE) {
       $replace = $node->toLink($node->label())->getUrl();
